### PR TITLE
kv: add DeadlockTimeout option to BatchRequest

### PR DESCRIPTION
**Pull request by iskettaneh**
This PR introduces a new `deadlock_timeout` field to the BatchRequest
Header struct. If set, it will be used as the time to wait on a lock
before pushing the lock holder for deadlock detection. If not set,
the cluster-wide setting:
kv.lock_table.deadlock_detection_push_delay will be used as the timeout.

A future PR will make this configurable at a session granularity.

References: https://github.com/cockroachdb/cockroach/issues/116305

Epic: None

Release note: None

### CODE REVIEW

**Review by arulajmani — COMMENTED** [review]
Nice job, this looks great. I only have a few nits, but otherwise this should be good to go. 

PS: sorry for taking so long to review this.

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-:-O3eZTzwA_xTYT3qwL2P:b-dv42u5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh and @nvanbenschoten)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eVCosE6Y6WqWYP9w8):*
"before pushing the lock holder for deadlock detection".
___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eVK2d-ysmQnTHMxkB):*
super nit: typically, we use "Informs" or "References" when linking github issues. 
___
*[`pkg/kv/kvpb/api.proto` line 2966 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eVT3a74SCdgmWGdkX:-O3eVT3b0Lxlj4JUxDcH:bj5io9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvpb/api.proto#L2966)):*
> ```proto
>   WriteOptions write_options = 35;
> 
>   // deadlock_timeout specifies the maximum amount of time that the batch
> ```

nit: this proto file will generate code which reads `DeadlockTimeout`, so we should use that here instead of `deadlock_timeout`.
___
*[`pkg/kv/kvpb/api.proto` line 2966 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eVgYoAsU8xBy4uC7M:-O3eVgYoAsU8xBy4uC7N:bx9zcdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvpb/api.proto#L2966)):*
> ```proto
>   WriteOptions write_options = 35;
> 
>   // deadlock_timeout specifies the maximum amount of time that the batch
> ```

Let's reword this to say "the amount of time that a request will wait on a lock before pushing the lock holder transaction to detect deadlocks".
___
*[`pkg/kv/kvpb/api.proto` line 2972 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eVwNd0XLuC3ZLREyS:-O3eVwNd0XLuC3ZLREyT:bizscb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvpb/api.proto#L2972)):*
> ```proto
>   // This differs from kv.lock_table.deadlock_detection_push_delay in that it
>   // gets applied at a finer granularity, where
>   // kv.lock_table.deadlock_detection_push_delay is a cluster-wide config.
> ```

We typically call these "cluster settings" instead of configs. Let's reword this to "a cluster wide setting".
___
*[`pkg/kv/kvpb/api.proto` line 2974 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eWJl90LQip1r7Gx81:-O3eWJl90LQip1r7Gx82:b-act2wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvpb/api.proto#L2974)):*
> ```proto
>   // kv.lock_table.deadlock_detection_push_delay is a cluster-wide config.
>   //
>   // A value of zero will cause the timeout duration to be taken from the
> ```

Consider rewording this paragraph to something like: "DeadlockTimeout, if set, takes precedence over the `kv.lock_table.deadlock_detection_push_delay` cluster setting. If unset, the cluster setting is used instead".

Separately, should this paragraph come before we talk about the difference between the cluster setting?
___
*[`pkg/kv/kvserver/concurrency/concurrency_control.go` line 441 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eXD_S8WSHmBAitLqF:-O3eXD_S8WSHmBAitLqG:b635j24) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvserver/concurrency/concurrency_control.go#L441)):*
> ```Go
> 	BaFmt redact.SafeFormatter
> 
> 	// The maximum amount of time that the batch request will wait on a lock
> ```

nit: DeadlockTimeout is the amount of time that the request will wait on a lock before pushing the lock holder's transaction for deadlock detection.
___
*[`pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks` line 1169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eYVIC6mg7deNzUv7_:-O3eYVIC6mg7deNzUv7a:b1940wv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks#L1169)):*
> ```
> # Class: lock holder aborted while pushing lock holder.
> #
> # Setup: txn1, txn2, txn3 acquire locks a, b, c
> ```

Can we use a simpler deadlock here? I was thinking one with just two transactions, txn1 and txn2. We can have txn1 lock key a, txn2 lock key b, have txn1 wait to acquire a lock on key b. Then, to create a deadlock, txn2 could try and acquire a lock on key a.
___
*[`pkg/kv/kvserver/concurrency/lock_table_waiter_test.go` line 195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eZ7_-0UJZUA5WiDC3:-O3eZ7_-0UJZUA5WiDC4:b-zdffqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvserver/concurrency/lock_table_waiter_test.go#L195)):*
> ```Go
> 			testutils.RunTrueAndFalse(t, "deadlock_timeout", func(t *testing.T, deadlock_timeout bool) {
> 				if deadlock_timeout {
> 					request.DeadlockTimeout = time.Nanosecond * 1
> ```

Why this change? `setupLockTableWaiterTest` manually overrides `LockTableDeadlockOrLivenessDetectionPushDelay` to be zero (to run fast), so this change doesn't seem in the spirit of this test.

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127430)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @nvanbenschoten)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eVCosE6Y6WqWYP9w8):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

"before pushing the lock holder for deadlock detection".
</blockquote></details>

Done.
___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eVK2d-ysmQnTHMxkB):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

super nit: typically, we use "Informs" or "References" when linking github issues.
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/api.proto` line 2966 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eVgYoAsU8xBy4uC7M:-O3h04kR0p5KQYjfdDTB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvpb/api.proto#L2966)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's reword this to say "the amount of time that a request will wait on a lock before pushing the lock holder transaction to detect deadlocks".
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/api.proto` line 2972 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eVwNd0XLuC3ZLREyS:-O3h0AlO6qqv1xZrOh_W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvpb/api.proto#L2972)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We typically call these "cluster settings" instead of configs. Let's reword this to "a cluster wide setting".
</blockquote></details>

Done.
___
*[`pkg/kv/kvpb/api.proto` line 2974 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eWJl90LQip1r7Gx81:-O3h0U15AL6ZBKk4csrE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvpb/api.proto#L2974)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Consider rewording this paragraph to something like: "DeadlockTimeout, if set, takes precedence over the `kv.lock_table.deadlock_detection_push_delay` cluster setting. If unset, the cluster setting is used instead".

Separately, should this paragraph come before we talk about the difference between the cluster setting?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks` line 1169 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eYVIC6mg7deNzUv7_:-O3h5U4jEyANoEXIdq3C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks#L1169)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Can we use a simpler deadlock here? I was thinking one with just two transactions, txn1 and txn2. We can have txn1 lock key a, txn2 lock key b, have txn1 wait to acquire a lock on key b. Then, to create a deadlock, txn2 could try and acquire a lock on key a.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/concurrency/lock_table_waiter_test.go` line 195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3eZ7_-0UJZUA5WiDC3:-O3h64Nr8qsNDv6y-aqd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7d86d317b91af52e296e79b45e8ce196dbf7047/pkg/kv/kvserver/concurrency/lock_table_waiter_test.go#L195)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Why this change? `setupLockTableWaiterTest` manually overrides `LockTableDeadlockOrLivenessDetectionPushDelay` to be zero (to run fast), so this change doesn't seem in the spirit of this test.
</blockquote></details>

Done.

**Review by arulajmani — APPROVED** [review]
:lgtm: modulo comment.

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-:-O3iUi7V6E_SSVg6Yytu:b-913mt2)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @iskettaneh and @nvanbenschoten)

___
*[`pkg/kv/kvpb/api.proto` line 2966 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3iUmRx2E_PIYTZaYU2:-O3iUmRx2E_PIYTZaYU3:b-p7zrjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6500e679d7b59b4b0811bb8fa751a6a10a94e760/pkg/kv/kvpb/api.proto#L2966)):*
> ```proto
>   WriteOptions write_options = 35;
> 
>   // DeadlockTimeout specifies the maximum amount of time that a request will
> ```

nit: let's drop the "maximum" here. 
___
*[`pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks` line 1173 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3iUBnR37m2D2x8CHfl:-O3iUBnS9XRhXzntbed2:b-nutoid) ([raw file](https://github.com/cockroachdb/cockroach/blob/6500e679d7b59b4b0811bb8fa751a6a10a94e760/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks#L1173)):*
> ```
> # Test:  set the cluster's deadlock timeout to 1h (debug-disable-txn-pushes)
> #        set the requests' deadlock timeout to 1ns (deadlock-timeout=1ns)
> #        txn1, txn2 read b, a
> ```

Instead of making these non-locking reads, let's have them acquire a lock instead (or be writes, which acquire locks with strength `Intent`). Who knows, maybe we change the non-locking reads blocking on writes behaviour in the future ;)

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127430)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani and @nvanbenschoten)

___
*[`pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks` line 1173 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/127430#-O3iUBnR37m2D2x8CHfl:-O3iaLry8syXzOmtps7g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6500e679d7b59b4b0811bb8fa751a6a10a94e760/pkg/kv/kvserver/concurrency/testdata/concurrency_manager/deadlocks#L1173)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Instead of making these non-locking reads, let's have them acquire a lock instead (or be writes, which acquire locks with strength `Intent`). Who knows, maybe we change the non-locking reads blocking on writes behaviour in the future ;)
</blockquote></details>

Done.
